src/Hamiltonians/TwoBodyRealCorrelation.jl

```diff
@@ -0,0 +1,65 @@
+"""
+    G2RealCorrelator(d::Int) <: AbstractHamiltonian{Float64}
+
+Two-body operator for density-density correlation between sites separated by `d`.
+```math
+    G_2(d) = \\sum_i^M \\hat{n}_i (\\hat{n}_{i+d} - \\delta_{0d}).
+```
+Can be applied to any [`SingleComponentFockAddress`](@ref).
+Assumes periodic boundary conditions where
+```math
+    G_2(-M/2 \\leq d < 0) = G_2(|d|),
+```
+```math
+    G_2(\\leq d > M/2) = G_2(M - d),
+```
+and normalisation
+```math
+    \\sum_{d=0}^{M-1} G_2(d) = N (N-1).
+```
+
+# Arguments
+- `d::Integer`: distance between sites.
+
+# See also
+
+* [`HubbardReal1D`](@ref)
+* [`G2Correlator`](@ref)
+* [`AbstractHamiltonian`](@ref)
+"""
+struct G2RealCorrelator{D} <: AbstractHamiltonian{Float64}
+end
+
+function G2RealCorrelator(d)
+    d::Int
+    return G2RealCorrelator{d}()
+end
```

Comment:
While I am sure that this code runs and is fit for the current purpose, I'm wondering what we should do with operators that are (so far) only used as observables to go into `AllOverlaps`.

For one thing, we subtype to `AbstractHamiltonian`, but the full interface is not implemented (in particular the `starting_address`). While this is debatable, it would be nice to be consistent with requiring an interface when subtyping. Eg. in automatic testing, we could just test the interface for all `AbstractHamiltonian`s.  A minor issue is the hard-wiring of the `Float64` type.

One solution of the missing `starting_address` would be to have a hierarchy like `AbstractHamiltonian{T} <: Operator{T}` and have this just be an `Operator`, or even a `Observable{T} <: Operator{T}`.

Another possibility would be to add the `starting_address`
```
struct G2RealCorrelator{T, D, A} :< AbstractHamiltonian{T}
    starting_address::A
end
```
where `T` could default to `Float64` (or `Int` for that matter) and the starting address could be provided by passing a related Hamiltonian. Having a starting_address has the advantage that `G2RealCorrelator` would already know which type of addresses will be thrown at it. Passing the Hamiltonian to a constructor (or constructing function) also would allow us to merge the current `G2Correlator` with `G2RealCorrelator` as information about whether we are in real or momentum space will be available.